### PR TITLE
Übung 04

### PR DESCRIPTION
**Pull request by GinaKastner**
Hab bisschen was an der Template und den Key Codes verändert, sodass man jetzt E/Eingabe verwenden kann um ein Spiel auszuwählen/neu zu starten und Q um bei FallingStones das laufende Spiel zu beenden und den Score zu sehen - hoffe geht in Ordnung, ansonsten kann ichs auch einfach zurücksetzen auf die ursprünglichen Einstellungen ^^

### CODE REVIEW

File: 04-Games/src/js/Inputs.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+export const Inputs = {
+    "primary": {
+        "keys": [69, 13], //Added key codes for E and Enter.
```
Comment (Neon-Ape) [inline 04-Games/src/js/Inputs.js:3]: Ja klar keycode für enter ist 13.
Aber warum funktioniert der keycode für E (groß e) für e und E und der keycode für e garnicht (das war nml der keycode den ich drin hatte und der aus unerfindlichen gründen nicht ging)

Ich hatte ja gehofft deine Änderungen bringen mir da etwas Verständnis, ich bin grad noch verwirrter.

File: 04-Games/src/js/games/GameTemplate.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    start() {}
+
+    bindControls() {}
+
+    tick(ctx) {
+        this.update(ctx);
+        if(this.gameOver || this.gameExit) {
+            this.gameOverScreen(ctx);
+            return;
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/GameTemplate.js:30]: Solche Änderungen sind nicht ok. 
Du änderst hier grundlegend wie sich die API im gameover verhält. 
In Pong führt das z.B. zu Problemen.
Vor allem ist es nicht nötig. Du kannst diese Funktionalität einfach in deiner erbenden Klasse überschreiben.

Comment (Neon-Ape) [reply 04-Games/src/js/games/GameTemplate.js:30]: Damit läuft ja auch die komplette Gamelogik im Hintergrund weiter, nur ohne dass du sie anzeigst.

File: 04-Games/src/js/games/FallingStones.js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+
+export class Bullet extends MovableGameObject {
+    
+    constructor(player, bulletSpeed) {
+        super(player.x + player.width/2, player.y, 10, 10, "#6bd26b", player.vx, -bulletSpeed);
+    }
+
+    bulletBorderPassed(ctx) {
+        return this.x < 0 || this.x > ctx.canvas.width || this.y < 0 || this.y > ctx.canvas.height;
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js]: Es ist üblich den warscheinlichsten Fall als erstes zu prüfen um schnellere returns zu bekommen.
In 90% der Fälle schießt man direkt nach oben, also y < 0 prüfen.
y > canvas.height kann theoretisch und praktisch nicht eintreten da wir nur nach oben schießen.

Der zusätzliche test auf x erzeugt in 100% der Fälle 200% zusätzlichen Rechenaufwand um einen recht kleinen edge case zu optimieren. 
Ja, du sparst dir für ein paar frames den collisioncheck mit bullets, die die canvas schon in x aber nicht y koordinate verlassen haben. Aber dafür machst du für jede Bullet jeden Frame einen Test der 50% der Rechenleistung einer Kollisionsabfrage braucht.

File: 04-Games/src/js/games/FallingStones.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        }
+    }
+
+    static get NAME() {
+        return "Falling Stones";
+    }   
+}
+
+
+export class Player extends MovableGameObject {
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js]: Hier hast du de-facto Paddle nachgebaut.
Könntest du dir auch sparen. Oder eben Paddle extenden wenn du den Konstruktor vereinfachen willst.

File: 04-Games/src/js/games/FallingStones.js [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+                if(GameObject.rectangleCollision(this.stones[i], this.bullets[j])) {
+                    this.points += 1;
+                    this.deleteStone(i);                    
+                    this.deleteBullet(j);
+                }
+            }
+        }
+    }
+
+    gameOverMessage() {
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js:124]: Die API-brechende Änderung an der Template mal dahingestellt, ist das eine echt umständliche Lösung etwas anderen Text darzustellen.
Du könntest genau so gut auf "secondary" klick den gameover text ändern.
Dann sparst du dir dass die komplette Gamelogik weiter läuft während du nur statischen text anzeigen willst.

Ich versteh auch nicht ganz warum es das überhaupt braucht. gameExit ändert außer dem Text nichts am game state.

File: 04-Games/src/js/games/FallingStones.js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+                }
+            }
+            //Inactive stones get deleted when out of screen.
+            else if(this.stones[i].stoneBorderPassed(ctx)) {
+                this.deleteStone(i);
+            }
+        }
+
+        //Hit
+        for(let i = 0; i < this.stones.length; i++) {
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js]: Das ist die dritte Schleife in der funktion die über stones läuft.
Das sollte alles in einer schleife sein, du verdreifachst hier grundlos den rechenaufwand.
Und da wir das alle 16ms (bei 60fps v-sync) ausführen ist der Unterschied relevant.
Du bist zwar noch ein Stück davon weg dass du die 60fps nicht halten kannst, aber gewöhn dir das lieber gleich ab. 
Sobald die schleifen über mehr als 3 Elemente iterieren merkst du den Unterschied auf einmal sehr.

File: 04-Games/src/js/games/FallingStones.js [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            if(this.stones[i].isActive) {
+                if(this.stones[i].stoneBorderReached(ctx)) {
+                    this.lives--;
+                    this.stones[i].isActive = false; //Inactive stone: Life already lost.
+                    if(this.lives === 0) this.gameOver = true;
+                }
+            }
+            //Inactive stones get deleted when out of screen.
+            else if(this.stones[i].stoneBorderPassed(ctx)) {
+                this.deleteStone(i);
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js:110]: Du machst dir hier einiges an Aufwand dadurch dass du leben schon abziehst wenn der Stein den Rand berührt.

File: 04-Games/src/js/games/FallingStones.js [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+    checkBullets(ctx) {
+        //Update
+        for(let i = 0; i < this.bullets.length; i++) {
+            this.bullets[i].update(ctx);
+        }
+        
+        
+        //Border
+        for(let i = 0; i < this.bullets.length; i++) {
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js]: Zwei for schleifen die das selbe machen.
Sollten zusammengefasst werden.

File: 04-Games/src/js/games/FallingStones.js [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    
+    generateStone(ctx) {
+        // Add new stone when 
+        // 1) there is no stone yet 
+        // 2) the last added stone has passed a certain percentage of the screen.
+        if(this.stones.length == 0 || (this.stones.length != 0 && this.stones[this.stones.length - 1].y >= ctx.canvas.height * this.stoneSpawnModifier)) { 
+            let positionStone;
+            do { 
+                positionStone = Math.random() * ctx.canvas.width;
+            } while (positionStone < 0 || positionStone > ctx.canvas.width - 50); //Ensure stone fully shown in screen (stone width = 50).
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js]: Du kannst dir die schleife sparen wenn du einfach Math.random() * width - 50 rechnest.

File: 04-Games/src/js/games/FallingStones.js [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+    deleteBullet(index) {
+        this.bullets.splice(index, 1);
+    }
+    
+    generateStone(ctx) {
+        // Add new stone when 
+        // 1) there is no stone yet 
+        // 2) the last added stone has passed a certain percentage of the screen.
+        if(this.stones.length == 0 || (this.stones.length != 0 && this.stones[this.stones.length - 1].y >= ctx.canvas.height * this.stoneSpawnModifier)) {
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js]: `this.stones.length != 0` ist an der stelle immer true. Der invertierte Fall ist ja vorher schon getestet,
Logischer operatoren brechen ab sobald sie einmal true sehen. Wenn also length 0 ist kommst du eh nicht in den hinteren check und muss somit auch nicht sicher stellen dass du das nur überprüfst wenn du != 0 bist.

File: 04-Games/src/js/games/FallingStones.js [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+    displayLives(ctx) {
+        ctx.fillStyle = "#AEEDBD";
+        ctx.font = "40px Arial";
+        ctx.textAlign = "center";
+        ctx.fillText(this.lives, this.player.x + this.player.width/2, this.player.y + this.player.height - 10);
+    }
+
+    
+    generateBullet() {
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js]: Du erstellst auf keydown und keyup eine Bullet. Also pro tastendruck 2.
Hier solltest du die übergebene boolean überprüfen um sicherzustellen dass du nur auf keydown schießt.

File: 04-Games/src/js/games/FallingStones.js [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+            }
+        }
+
+        //Hit
+        for(let i = 0; i < this.stones.length; i++) {
+            for(let j = 0; j < this.bullets.length; j++) {
+                if(GameObject.rectangleCollision(this.stones[i], this.bullets[j])) {
+                    this.points += 1;
+                    this.deleteStone(i);                    
+                    this.deleteBullet(j);
```
Comment (Neon-Ape) [inline 04-Games/src/js/games/FallingStones.js:118]: Du löschst Elemente der Arrays über die du inkrementierend iterierst.
Das wirft dir immer wenn du ein Element löschst einen Error, weil die indizes nicht mehr stimmen.
In solchen Fällen immer dekrementierende (i--) Schleifen verwenden.

File: 04-Games/src/js/games/FallingStones.js (hunk 5 above)
Comment (Neon-Ape) [reply 04-Games/src/js/games/FallingStones.js:124]: Ich kann halt danach noch weiterspielen und mehr Punkte sammeln, ich seh halt nur das Spiel nicht. Dafür bin ich unsterblich.
Würd ich jetzt schon nen game-breaking bug nennen.

**Review by Neon-Ape — CHANGES_REQUESTED** [review]
Von der Spielbarkeit eine der schönsten Abgaben.

Allerdings hast du dir mit dem Überschreiben der GameTemplate einen bösen Bug geangelt und ähnliche Bugs in allen anderen Spielen ausgelöst.
Grundsätzlich musst du auch mehr auf performance achten. Wenn der komplette Code maximal 16ms dauern darf sollte man sich angewöhnen unnötigen Code zu vermeiden.

Ich musste deine Änderungen auch mehr suchen als nötig weil der Merge teil des pull requests ist. Da fehlt dann das change highlighting weil logischerweise alles neu ist.

File: 04-Games/src/js/Inputs.js (hunk 1 above)
Comment (GinaKastner) [reply 04-Games/src/js/Inputs.js:3]: Hab mir das eigtl. auch nur aus der Tabelle hier rausgesucht, lt. der wäre nämlich 101 der Code für Numpad 5 und 113 f2?
https://keycode.info

Comment (GinaKastner) [reply 04-Games/src/js/Inputs.js:3]: Btw. sorry, dass du die Änderungen suchen musstest, damit es mit dem Merge funktioniert, hätte ich dann vermutlich zuerst den master mit dem Starter für Ü4 updaten müssen oder?

File: 04-Games/src/js/games/FallingStones.js (hunk 4 above)
Comment (GinaKastner) [reply 04-Games/src/js/games/FallingStones.js]: Gedanke dahinter war, dass Paddle ja nicht allgemein in GameObjects sondern in Pong enthalten ist und ich daher davon ausgegangen bin, dass es eine spielspezifische Klasse sein soll, die nur innerhalb von Pong verwendet/erweitert werden soll.

File: 04-Games/src/js/games/GameTemplate.js (hunk 2 above)
Comment (GinaKastner) [reply 04-Games/src/js/games/GameTemplate.js:30]: Dass das Spiel dadurch weiterläuft hab ich nicht bedacht, überleg mir wie ich das anders umsetzen kann ^^'

File: 04-Games/src/js/games/FallingStones.js (hunk 9 above)
Comment (GinaKastner) [reply 04-Games/src/js/games/FallingStones.js]: Wenn ich es mit deiner Lösung mache, werden aber auch teilweise vom Bildschirmrand abgeschnittene Steine generiert, was ich mit der Schleife vermeiden wollte:
Screenshot (159)_LI

File: 04-Games/src/js/games/FallingStones.js (hunk 11 above)
Comment (GinaKastner) [reply 04-Games/src/js/games/FallingStones.js]: Wie das funktioniert ist mir nicht ganz klar, welchen boolean meinst du?

File: 04-Games/src/js/games/FallingStones.js (hunk 9 above)
Comment (Neon-Ape) [reply 04-Games/src/js/games/FallingStones.js]: Das liegt glaub ich am padding im css. 
Da kannst du aber auch den minimalwert anheben um das auszuschließen.
Math.random() * (width - (50 + 10)) + 10 
würde jetzt zB. 10px einrücken an beiden rändern.

File: 04-Games/src/js/games/FallingStones.js (hunk 4 above)
Comment (Neon-Ape) [reply 04-Games/src/js/games/FallingStones.js]: Code wiederbenutzen wo es geht ;)

File: 04-Games/src/js/Inputs.js (hunk 1 above)
Comment (Neon-Ape) [reply 04-Games/src/js/Inputs.js:3]: Ja dass verwirrt mich daran so fest, wenn ich einfach nen listener setze und dann "e" drücke bekomm ich diese Ausgabe. 😕 
image
Wie ich sehe hat sich aber mittlerweile scheinbar .key über .which durchgesetzt. Js ist halt auch immer im Wandel.
Weirdness intensifies: wenn ich den listener auf keydown setzte bekomm ich `69`... 
Naja man lernt mal wieder dass man mit dem Event dass man auch hernehmen will testen sollte, nicht mit einem von dem man ausgeht dass es sich identisch verhält.

File: 04-Games/src/js/games/FallingStones.js (hunk 9 above)
Comment (GinaKastner) [reply 04-Games/src/js/games/FallingStones.js]: Ah okay, danke!
